// wallet/index.js
const { exec } = require('child_process');

function runCommand(cmd) {
  return new Promise((resolve, reject) => {
    exec(cmd, { maxBuffer: 10 * 1024 * 1024 }, (err, stdout, stderr) => {
      if (err) return reject(stderr || err.message);
      try {
        resolve(JSON.parse(stdout));
      } catch (e) {
        resolve({ raw: stdout.trim() });
      }
    });
  });
}

const api = {
  // ---------------- Bitcoin Core ----------------
  async startBitcoin() {
    return runCommand(`bitcoind -regtest \
      -daemon \
      -fallbackfee=0.0002 \
      -rpcuser=foo \
      -rpcpassword=bar \
      -rpcallowip=127.0.0.1 \
      -txindex`);
  },

  async stopBitcoin() {
    return runCommand('bitcoin-cli -regtest -rpcuser=foo -rpcpassword=bar stop');
  },

  async getInfoBitcoin() {
    return runCommand('bitcoin-cli -regtest -rpcuser=foo -rpcpassword=bar getblockchaininfo');
  },

  async newBitcoinAddress() {
    return runCommand('bitcoin-cli -regtest -rpcuser=foo -rpcpassword=bar getnewaddress');
  },

  async sendBitcoin(address, amount) {
    return runCommand(`bitcoin-cli -regtest -rpcuser=foo -rpcpassword=bar sendtoaddress ${address} ${amount}`);
  },

  async fundLightningNode(address, blocks) {
    return runCommand(`bitcoin-cli -regtest -rpcuser=foo -rpcpassword=bar generatetoaddress ${blocks} ${address}`);
  },

  // ---------------- Lightning Node 1 ----------------
  async startLightning() {
    return runCommand(`lightningd --network=regtest \
      --lightning-dir=/Users/a1707/lightning \
      --addr=127.0.0.1:9735 \
      --bitcoin-rpcuser=foo \
      --bitcoin-rpcpassword=bar \
      --bitcoin-rpcconnect=127.0.0.1 \
      --bitcoin-rpcport=18443 \
      --disable-plugin=clnrest \
      --disable-plugin=wss-proxy \
      --log-level=debug \
      --log-file=/Users/a1707/lightning/debug.log \
      --daemon`);
  },

  async stopLightning() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning stop');
  },

  async getInfoLightning() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning getinfo');
  },

  async newLightningAddress() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning newaddr');
  },

  async listLightningFunds() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning listfunds');
  },

  async connectPeer(peerId, host, port) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning connect ${peerId}@${host}:${port}`);
  },

  async listPeers() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning listpeers');
  },

  async openChannel(peerId, satoshis) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning fundchannel ${peerId} ${satoshis}`);
  },

  async createInvoiceNode1(amountSats, label = "invoice-node1", description = "Test invoice from Node1") {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning invoice ${amountSats} ${label}-${Date.now()} "${description}"`);
  },

  async payInvoiceNode1(bolt11) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning pay ${bolt11}`);
  },

  async listInvoicesNode1() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning listinvoices');
  },

  async listPaysNode1() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning listpays');
  },

  // ---------------- Lightning Node 2 ----------------
  async startLightning2() {
    return runCommand(`lightningd --network=regtest \
      --lightning-dir=/Users/a1707/lightning2 \
      --addr=127.0.0.1:9736 \
      --bitcoin-rpcuser=foo \
      --bitcoin-rpcpassword=bar \
      --bitcoin-rpcconnect=127.0.0.1 \
      --bitcoin-rpcport=18443 \
      --disable-plugin=clnrest \
      --disable-plugin=wss-proxy \
      --log-level=debug \
      --log-file=/Users/a1707/lightning2/debug.log \
      --daemon`);
  },

  async stopLightning2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 stop');
  },

  async getInfoLightning2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 getinfo');
  },

  async newLightningAddress2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 newaddr');
  },

  async listLightningFunds2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 listfunds');
  },

  async listPeers2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 listpeers');
  },

  async openChannel2(peerId, satoshis) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 fundchannel ${peerId} ${satoshis}`);
  },

  async createInvoiceNode2(amountSats, label = "invoice-node2", description = "Test invoice from Node2") {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 invoice ${amountSats} ${label}-${Date.now()} "${description}"`);
  },

  async payInvoiceNode2(bolt11) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 pay ${bolt11}`);
  },

  async listInvoicesNode2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 listinvoices');
  },

  async listPaysNode2() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 listpays');
  },

  // ---------------- Lightning Node 3 (Initial Setup Only) ----------------
  async startLightning3() {
    return runCommand(`lightningd --network=regtest \
      --lightning-dir=/Users/a1707/lightning3 \
      --addr=127.0.0.1:9737 \
      --bitcoin-rpcuser=foo \
      --bitcoin-rpcpassword=bar \
      --bitcoin-rpcconnect=127.0.0.1 \
      --bitcoin-rpcport=18443 \
      --disable-plugin=clnrest \
      --disable-plugin=wss-proxy \
      --log-level=debug \
      --log-file=/Users/a1707/lightning3/debug.log \
      --daemon`);
  },

  async stopLightning3() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 stop');
  },

  async getInfoLightning3() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 getinfo');
  },

  async newLightningAddress3() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 newaddr');
  },

  async listLightningFunds3() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 listfunds');
  },

  async listPeers3() {
    return runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 listpeers');
  },

  async connectPeer3(peerId, host, port) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 connect ${peerId}@${host}:${port}`);
  },

  async openChannel3(peerId, satoshis) {
    return runCommand(`lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 fundchannel ${peerId} ${satoshis}`);
  },

  // ---------------- Shared Helpers ----------------
  async listChannels() {
    const node1 = await runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning listchannels');
    const node2 = await runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning2 listchannels');
    const node3 = await runCommand('lightning-cli --network=regtest --lightning-dir=/Users/a1707/lightning3 listchannels');
    return { node1, node2, node3 };
  },

  async verifyChannelFunds() {
    const node1Funds = await this.listLightningFunds();
    const node2Funds = await this.listLightningFunds2();
    const node3Funds = await this.listLightningFunds3();
    return { node1Funds, node2Funds, node3Funds };
  }
};

module.exports = api;
